Fixes for avoidance of hosts taking backup in PRS & ERS

ejortegau opened the pull request:
<!--
  Thank you for your contribution to the Vitess project.
  How to contribute: https://vitess.io/docs/contributing/
  Please first make sure there is an open Issue to discuss the feature/fix suggested in this PR.
  If this is a new feature, please mark the Issue as "RFC".
 -->

<!-- if this PR is Work in Progress please create it as a Draft Pull Request -->

## Description
This PR addresses fixes issues in PRS & ERS preference for tablets that are not taking backups. It does away with  redundant field definitions in some of the `proto` messages and addresses segfault that could take place during ERS.
 
## Related Issue(s)

https://github.com/vitessio/vitess/issues/17299
#16997 

## Checklist

-   [x] "Backport to:" labels have been added if this change should be back-ported to release branches
-   [x] If this change is to be back-ported to previous releases, a justification is included in the PR description
-   [x] Tests were added or are not required
-   [x] Did the new or modified tests pass consistently locally and on CI?
-   [x] Documentation was added or is not required

## Deployment Notes

N/A

ejortegau commented on go/vt/vttablet/grpctmserver/server.go:
```diff
@@ -354,7 +354,7 @@ func (s *server) MysqlHostMetrics(ctx context.Context, request *tabletmanagerdat
 func (s *server) ReplicationStatus(ctx context.Context, request *tabletmanagerdatapb.ReplicationStatusRequest) (response *tabletmanagerdatapb.ReplicationStatusResponse, err error) {
 	defer s.tm.HandleRPCPanic(ctx, "ReplicationStatus", request, response, false /*verbose*/, &err)
 	ctx = callinfo.GRPCCallInfo(ctx)
-	response = &tabletmanagerdatapb.ReplicationStatusResponse{BackupRunning: s.tm.IsBackupRunning()}
```
This information is already being set in `rpc_replication.go`

ejortegau commented on go/vt/vttablet/grpctmserver/server.go:
```diff
@@ -638,8 +638,6 @@ func (s *server) StopReplicationAndGetStatus(ctx context.Context, request *table
 		response.Status = statusResponse.Status
 	}
 
-	response.BackupRunning = s.tm.IsBackupRunning()
```
This information is already being set in `rpc_replication.go`

ejortegau commented on proto/replicationdata.proto:
```diff
@@ -66,7 +66,6 @@ message Configuration {
 message StopReplicationStatus {
   replicationdata.Status before = 1;
   replicationdata.Status after = 2;
-  bool backup_running = 3;
```
A field for this is already present in replicationdata.Status, no need to add another one here which just led to confusion in the previous PR.

dbussink replied:
This would be a backwards compatibility breaking change. I don't know if we can safely remove it, but if we can, the first step would be to explicitly mark it as deprecated before it can be removed. We can't in general just delete proto fields without any explicit deprecation process.

deepthi replied:
The original PR that added this field has not gone into any release, so it would be ok to simply delete the field.

dbussink replied:
Ah, that's good then!

ejortegau replied:
This field was just added a couple of weeks ago and has not made it into any release yet. So I think that removing it is safe, but can also leave it in place marking it as deprecated if you prefer.

ejortegau commented on proto/tabletmanagerdata.proto:
```diff
@@ -360,7 +360,6 @@ message ReplicationStatusRequest {
 
 message ReplicationStatusResponse {
   replicationdata.Status status = 1;
-  bool backup_running = 2;
```
A field for this is already present in replicationdata.Status, no need to add another one here which just led to confusion in the previous PR.

ejortegau commented on proto/tabletmanagerdata.proto:
```diff
@@ -549,7 +548,6 @@ message StopReplicationAndGetStatusResponse {
 
   // Status represents the replication status call right before, and right after telling the replica to stop.
   replicationdata.StopReplicationStatus status = 2;
-  bool backup_running = 3;
```
A field for this is already present in replicationdata.Status, no need to add another one here which just led to confusion in the previous PR.

deepthi reviewed (approved):
LGTM

deepthi reviewed (changes requested):
While these changes look good, the fact that the previous PR caused vtctld to crash points to missing tests.
Can you come up with a test case that would reproduce that particular crash and add it into this PR please?

ejortegau commented:
> > While these changes look good, the fact that the previous PR caused vtctld to crash points to missing tests. Can you come up with a test case that would reproduce that particular crash and add it into this PR please?
> 
> I noticed that one of the commits says "fix SIGSEGV in test". Which test was that, and how was it passing on the previous PR?

The first PR accessed the flag indicating whether a backup was or not running directly from a field in a struct. The new code access it from a field in a pointer to the struct. Thus the previous PR always found something there (even if it was the default false), whereas the new one could get a nil struct and crash. Hence why that commit wrapped that in some checks for the struct pointer not being null.

ejortegau commented:
> While these changes look good, the fact that the previous PR caused vtctld to crash points to missing tests. Can you come up with a test case that would reproduce that particular crash and add it into this PR please?

The crash during ERS was a consequence of my first attempt to fix the fact that I was using the wrong value for figuring out whether the backup was running or not. By replacing [this](https://github.com/vitessio/vitess/pull/17300/commits/fc1faaebf32d65cb64c4117d8cc679e5694767a1#diff-6524ceca7ed94041330ed7147fa8a284e64b1c7cbbf5d12d72fede7ef9419196L219-L221) with [this](https://github.com/vitessio/vitess/pull/17300/commits/fc1faaebf32d65cb64c4117d8cc679e5694767a1#diff-6524ceca7ed94041330ed7147fa8a284e64b1c7cbbf5d12d72fede7ef9419196R242-R245) I introduced `stopReplicationStatus.After.BackupRunning` when `After` might be null. This didn't exist in the original PR and hence a corresponding test didn't exist or made no sense back then.

In any case, I just added some test cases to ensure that the value of the map with backup state information is correct, and to verify that if the `After` replication state is null, no segfault takes place and the value is taken from the before state.
